user/ddns3-client/ddns3.1
.TH DDNS3 1 "$Date: 2003/03/20 05:21:59 $"

.SH NAME
ddns3 \- ddns.nu remote update client

.SH SYNOPSIS
.B ddns3 --user
.I user
.B --pass
.I password
.B [
.I options
.B ]
.I command
.B ...

.SH DESCRIPTION
.B ddns3
updates IP-Handles on the ddns.nu remote update server.  This
allows remote automated updates rather than using the website
via a browser.

.SH OPTIONS
.IP "--user user"
Specifies the username to authenticate with.
.IP "--pass password"
Specifies the password to authenticate with.

.IP "--host hostname"
Override the default server hostname to connect to.
The default server hostname is 
.B ns.ddns.nu.
.IP "--port portnumber"
Override the default TCP port to connect to.
The default TCP port for the ddns3 protocol is
.B 2164
.IP "--auth authtype"
Specify what type of authentication protocol to use.
The available types are:
.B plaintext,
.B crypt,
.B md5,
.B ddns,
and
.B strong.
The default type is
.B ddns
which is secure in the sense that the username is sent in 
cleartext, but the password is sent hashed in such a manner
as to defeat replay attacks by utilising a salt sent from
the server.

.SH COMMANDS
.IP "list"
List all IP-Handles and their current values for the account.
You may do this multiple times, for example it is perfectly
legal (and perhaps even useful) to do a
.B list,
followed by a
.B set,
and then another
.B list
to observe the change.
.IP "set handle-name ip-value"
Set an IP-Handle named
.B handle-name
to the IP value
.B ip-value.
Similar to
.B list
you can specify as many such operations on the command line
as you wish and they will be executed in the order specified.
.IP "guess handle-name ip-source"
Set the IP-Handle named
.B handle-name
to the IP value derived from IP seen by the
.B ip-source
end of the update connection.
The 
.B ip-source
is specified as either
.B remote
or
.B local.
Again you may specify multiple
.B guess
commands on the command line and they will be executed in order.

Using the 
.B local
option saves you from having to specify the IP on the command line.
Local guessing usually gets the IP correct as it takes the local
address of the socket making the update connection.  With conventional
routing arrangements this is likely to be the IP you want.

However the
.B remote
option sets the IP-Handle to the IP that the server sees the connection
as originating from.  This is most useful if you are running the client
from behind a NAT gateway which itself is incapable of running the client.
In this case the IP of the NAT gateway is used, rather than the probably
martian IP of the host the client is actually executing on.

.SH DIAGNOSTICS
Useful error messages are produced on stderr, all other runtime output goes
to stdout.

On runtime exception the return value of the process may be useful in scripting:

-1
.RS
The usage message was generated by running the command with no arguments.
.RE
-2
.RS
The argument parser didn't like the command line.
.RE
-3
.RS
A connection could not be established with the update server.
.RE
-4
.RS
Authentication failed.
.RE
-5
.RS
Network errors closing the connection (these can often be ignored).
.RE
-6
.RS
Some kind of internal logic error in the process, like a malloc/free failure.
.RE

The process returns a positive count of the number of commands executed on
the server that generated an error, or zero on error free completion of all
commands.

.SH BUGS
The authentication details should be read from a file to avoid 
their visibility in the process list. There should be support 
for other ddns3 protocol primitives like
.B MOTD
and
.B HELP
for completeness.  The java implementation 
.B jddns3
does this, and is largely command line equivalent.  Short network
writes are unchecked in at least one place.

.SH AUTHOR
Alan Yates <alany@ay.com.au>

.SH VERSION
$Id: ddns3.1,v 1.2 2003/03/20 05:21:59 alany Exp $

.SH "SEE ALSO"
http://www.ddns.nu/

http://www.ddns.nu/files/ddns-v3.txt


freeswan/CHANGES
New and notable in 1.97:

Sanity check added for at least one of AH or ESP more for checking if
"make *config" has even been run.

Fixed 2.2 local IKE fragmentation blackhole.  Still won't work if
iptraf or another pcap app is running.

MCR added ikeping(8).

Applied DHR's tunnel patch to streamline IKE/specialSA processing.

Added local and remote IKE bypass tests to the UML test suite.

UML KLIPS %trap* tests now check for upbound pf_key messages.

An ICMP ICMP_DEST_UNREACH, ICMP_PKT_FILTERED is now sent upon %reject.

Added a UML input file with first and last packets being outside the 
covered range to "sandwich" the packets under test.

The default _updown script has been cleaned up slightly, reducing
internal duplication that arose accidentally.

When doing --add on an RSA-signature conn, ipsec_auto now tells Pluto
about the keys before, not after, telling it about the conn.  The main
consequence of this is that if something is wrong with one of the keys,
the conn itself does *not* get loaded.  This should reduce confusion.

The rp_filter checks during KLIPS startup have been fixed to be a bit less
paranoid; rp_filter is thought to cause problems only on the physical
interfaces, not on the ipsecN interfaces. 

The configuration process has been fixed to be a bit more paranoid.  Some
users have had mysterious compile failures with complaints involving names
like MD5_CTX; they seem to have resulted from kernel configuration changes
somehow getting into .config but not into include/linux/autoconf.h (which
is what actually controls most of the kernel compile).  The post-config
checking now verifies not only that .config has the right stuff in it, but
that autoconf.h does too.  (We still don't see *how* this discrepancy can
arise, but this will at least catch it when it does.)

The details of linking the KLIPS source into the kernel are now under
control of a configuration variable in Makefile.inc, to make it easier
for distribution builders to change.

The top-level Makefile now includes cleverness which assigns a date-based
version code if a build is done from a raw CVS checkout.

The library now includes a new function, ttodatav(), like ttodata() but
in some cases it gives more verbose error reports.  Also, there are now
a couple of functions which distill an RSA public key to a "key ID" for
use in messages, a crypto-quality pseudo-random-number generator, and a
function that converts a single address into a (singleton) subnet.

Barf's secret/key-censoring utilities have been tidied up so that what
they emit matches the key-ID library functions. 

The ipsec command now supplies a couple more environment variables to
things it calls, and this has reduced the number of things which need to
be edited at build time to know about configuration variables. 

The configuration reader now allows parameter names to begin with "_";
these are strictly reserved for internal development purposes and will
not in general be documented.


As usual, there are assorted small bug fixes and improvements to docs
and messages.



New and notable in 1.96:

Compatibility with the 2.0.xx kernels (under Red Hat 5.2) is broken.
This was accidental, and we have some hopes of fixing it, although it
won't last forever even if we do.  See README.

Our "IKE Implementation Issues" IETF draft has been added to our docs.

Newhostkey now has a mandatory --output "option" specifying a filename
(although the magic filename "-" means "standard output"), and sets
restrictive permissions on newly-created files to try to keep private keys
truly private. 

Pluto now gracefully handles a couple of odd error cases (missing private
keys, network errors at just the wrong time in negotiations) which used to
cause it to die, provoking an automatic restart of the IPsec subsystem. 

Makefile.inc now has more detailed control over where manpages go, to ease
integration into distributions.  The name of the destination directory for
RPMs is also now centralized there.

Rsasigkey now checks for certain obscure internal failures, which can
happen if --random is used to supply a source of random bits that aren't
really very random, and complains accordingly.

Barf's eroute table, formerly sorted by source, is now sorted by destination.
This is probably a bit clearer for hosts with lots of eroutes.

Use of the %reject target in a shunt eroute now, as long planned, actually
causes an ICMP rejection message (Destination Unreachable:  Communication
Administratively Prohibited). 

KLIPS now declares its module license to be "Dual BSD/GPL".  (It is 
dual because the libdes and zlib code is BSD licensed, while KLIPS itself 
is GPL).  Whether this is exactly right is not yet clear, but at least it
stops the latest module loaders from whining.

The UML test stuff continues to grow and add new test cases.

As usual, there are assorted small bug fixes and improvements to docs
and messages.



New and notable in 1.95:

The last remnants of the "%hold" bug, which broke 1.93 and 1.94, have
(we think) been dealt with.

Minor bugs in the RPM-building stuff have been fixed, and it has been
updated to deal with Red Hat's abrupt removal of the --buildroot option
from rpm.

Pluto now supports the ability to delete an instance of a conn, rather
than just the whole conn (for cases, notably OE, where a single connection
description can be involved in multiple connections).  There is no support
for this in ipsec_auto yet.

Pluto's routine (non-debug) logging is now somewhat terser, and some of
the more boring messages are produced only if --debug-lifecycle is turned
on (this is a temporary hack, it's not really debugging output).

Pluto's innards have been reorganized somewhat, in preparation for making
DNS lookups asynchronous.  There should be no user-visible effects.

Pluto now ignores IPv6 link-local interface addresses; we think that they
are never relevant, and trying to bind to them causes difficulties.  This

user/ddns3-client/ddns3.c
/*
 *	DDNS v3 Client
 *
 *		Author:		Alan Yates <alany@ay.com.au>
 *		Win32 Port:	Kieron Briggs <kieron@kieron.nu>
 *		Version:	$Id: ddns3.c,v 1.2 2003/03/20 05:21:59 alany Exp $
 */
#include <stdio.h>
#ifdef WIN32
#include <windows.h>
#define strcasecmp(x,y) stricmp((x),(y))
else
#include <unistd.h>
#endif
#include <string.h>
#include <stdlib.h>
#include "ctx.h"

#define	ERR_USAGE	-1
#define	ERR_ARGUMENTS	-2
#define	ERR_CONNECTION	-3
#define	ERR_AUTH	-4
#define ERR_CLOSING	-5
#define ERR_LOGIC	-6

struct client_vars {
	char *host;
	char *port;
	char *auth;
	char *user;
	char *pass;
	int debug;
} CV;

static void
help(void) {
	fprintf(stderr, "\n");
	fprintf(stderr, "DDNS Client v3 (c) :: http://www.ddns.nu/\n");
	fprintf(stderr, "(c) 1999-2002 Alan Yates <alany@ay.com.au>\n");
	fprintf(stderr, "$Id: ddns3.c,v 1.2 2003/03/20 05:21:59 alany Exp $\n");
	fprintf(stderr, "\n");
	fprintf(stderr, "Usage: ddns3 <required args> [optional args] <action list>\n");
	fprintf(stderr, "\n");
	fprintf(stderr, "Required args\n");
	fprintf(stderr, "	--user	username\n");
	fprintf(stderr, "	--pass	password\n");
	fprintf(stderr, "Optional args\n");
	fprintf(stderr, "	--host	server-hostname		specify server host\n");
	fprintf(stderr, "	--port	server-tcp-port		specify server port\n");
	fprintf(stderr, "	--debug				activate debug mode\n");
	fprintf(stderr, "	--auth	auth-name		specify authentication type\n");
	fprintf(stderr, "Authentication types:\n");
#ifdef WIN32
	fprintf(stderr, "	plaintext, md5, ddns (default), strong\n");
#else
	fprintf(stderr, "	plaintext, crypt, md5, ddns (default), strong\n");
#endif
	fprintf(stderr, "Actions:\n");
	fprintf(stderr, "	list				lists IP-Handles\n");
	fprintf(stderr, "	set <handle> <ip>		sets an IP-Handle's value\n");
	fprintf(stderr, "	guess <handle> <source>		sets an IP-Handle's value\n");
	fprintf(stderr, "					guessing the IP from the <source>\n");
	fprintf(stderr, "					end of the connection, ends are\n");
	fprintf(stderr, "					named 'remote' or 'local'\n");
	fprintf(stderr, "\n");
}

static int
parse_args(int argc, char **argv, char **env) {
	int i;

	if(argc == 1) {
		help();
		exit(ERR_USAGE);
	}

	for(i = 1; i < argc; i++) {
		if(!strcmp("--user", argv[i]) && i+1 < argc) {
			CV.user = argv[++i];
		} else if(!strcmp("--pass", argv[i]) && i+1 < argc) {
			CV.pass = argv[++i];
		} else if(!strcmp("--host", argv[i]) && i+1 < argc) {
			CV.host = argv[++i];
		} else if(!strcmp("--port", argv[i]) && i+1 < argc) {
			CV.port = argv[++i];
		} else if(!strcmp("--auth", argv[i]) && i+1 < argc) {
			CV.auth = argv[++i];
		} else if(!strcmp("--debug", argv[i])) {
			CV.debug = -1;
		}
		else if(!strcmp("list", argv[i])) break;
		else if(!strcmp("set", argv[i])) break;
		else if(!strcmp("guess", argv[i])) break;
		else {
			fprintf(stderr, "unknown or malformed argument: %s\n", argv[i]);
			exit(ERR_ARGUMENTS);
		}
	}
	return i;
}

static void
defaults(void) {
	CV.host = "ns.ddns.nu";
	CV.port = "5000";
	CV.auth = "ddns";
	CV.debug = 0;
}

static void
check_args(void) {
	int err = 0;

	if(!CV.user) {
		fprintf(stderr, "client_error: you must specify a username\n");
		err ++;
	}

	if(!CV.pass) {
		fprintf(stderr, "client_error: you must specify a password\n");
		err ++;
	}

	if(err > 0) {
		fprintf(stderr, "client_exit: argument parsing; %d errors\n", err);
		exit(ERR_ARGUMENTS);
	}
}

static int
actions(struct ddns3_ctx *c, int argc, char **argv, int start) {
	int i, err = 0;
	for(i = start; i < argc; i++) {
		if(!strcmp("list", argv[i])) {
			if(ddns3_ctx_list(c) < 0) {
				fprintf(stderr, "server_message: %s", c->buf);
				fprintf(stderr, "client_error: could not get handle list\n");
				err++;
			} else {
				int j;
				for(j = 0; c->list[j]; j += 2) {
					printf("%-20s %s\n", c->list[j], c->list[j+1]);
				}
			}
		} else if(!strcmp("set", argv[i]) && i+2 < argc) {
			if(ddns3_ctx_set(c, argv[i+1], argv[i+2]) < 0) {
				fprintf(stderr, "server_message: %s", c->buf);
				fprintf(stderr, "client_error: could not set handle %s to ip %s, closing\n", argv[i+1], argv[i+2]);
				err++;
			} else printf("server_message: %s", c->buf);
			
			i += 2;
		} else if(!strcmp("guess", argv[i]) && i+2 < argc) {
			int type = -1;
			if(!strcasecmp("remote", argv[i+2])) type = GUESS_REMOTE;
			else if(!strcasecmp("local", argv[i+2])) type = GUESS_LOCAL;
			else {
				fprintf(stderr, "client_error: invalid guess type: %s\n", argv[i+2]);
				return ERR_ARGUMENTS;
			}
			if(ddns3_ctx_guess(c, argv[i+1], type) < 0) {
				fprintf(stderr, "server_message: %s", c->buf);
				fprintf(stderr, "client_error: could not set handle %s by guessing IP, closing\n", argv[i+1]);
				err++;
			} else printf("server_message: %s", c->buf);

			i += 2;
		} else {
			fprintf(stderr, "unknown or mangled action: %s\n", argv[i]);
			return ERR_ARGUMENTS;
		}
	}
	
	return err;
}

int
main(int argc, char **argv, char **env) {
	int start, ret;
	struct ddns3_ctx *c;

	/* arguments and ugly stuff */
	memset(&CV, 0, sizeof(struct client_vars));
	defaults();
	start = parse_args(argc, argv, env);
	check_args();

	/* now the long and ugly stuff */

	/* connect and login to DDNS server */
	if(ddns3_ctx_new(&c, CV.host, atoi(CV.port)) < 0) {
		fprintf(stderr, "client_error: creating ddns3_ctx, exiting\n");
		exit(ERR_LOGIC);
	}
	if(ddns3_ctx_connect(c) < 0) {
		fprintf(stderr, "server_message: %s", c->buf);
		fprintf(stderr, "client_error: could not connect to %s:%d, exiting\n", CV.host, atoi(CV.port));
		exit(ERR_CONNECTION);
	}
	if(ddns3_ctx_login(c, CV.auth, CV.user, CV.pass) < 0) {
		fprintf(stderr, "server_message: %s", c->buf);
		fprintf(stderr, "client_error: could not authenticate\n");
		exit(ERR_AUTH);
	}

	/* do what we have to */
	ret = actions(c, argc, argv, start);

	/* close connection nicely */
	if(ddns3_ctx_logout(c) < 0) {
		fprintf(stderr, "server_message: %s", c->buf);
		fprintf(stderr, "client_error: could not logout, giving up clean goodbye, exiting\n");
		exit(ERR_CLOSING);
	}
	if(ddns3_ctx_disconnect(c) < 0) {
		fprintf(stderr, "server_message: %s", c->buf);
		fprintf(stderr, "client_error: disconnect failed, giving up\n");
		exit(ERR_CLOSING);
	}
	if(ddns3_ctx_del(&c) < 0) {
		fprintf(stderr, "client_error: could not delete ddns3_ctx, oh well\n");
		exit(ERR_LOGIC);
	}
	
	return ret;
}
